contracts/asset-proxy/contracts/src/bridges/MixinGasToken.sol

```diff
@@ -0,0 +1,55 @@
+/*
+
+  Copyright 2019 ZeroEx Intl.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+*/
+
+pragma solidity ^0.5.16;
+
+import "@0x/contracts-utils/contracts/src/DeploymentConstants.sol";
+import "../interfaces/IGasToken.sol";
+
+
+contract MixinGasToken is
+    DeploymentConstants
+{
+
+    /// @dev Frees gas tokens based on the amount of gas consumed in the function
+    modifier freesGasTokens {
+        uint256 gasBefore = gasleft();
+        _;
+        IGasToken gst = IGasToken(_getGstAddress());
+        if (address(gst) != address(0)) {
+            // (gasUsed + FREE_BASE) / (2 * REIMBURSE - FREE_TOKEN)
+            //            14154             24000        6870
+            uint256 value = (gasBefore - gasleft() + 14154) / 41130;
```

Comment:
This tries to refund the entire amount of gas consumed, right? Maybe we should just try to refund half. I believe a quirk of the EVM is that refunds are capped at half the gas consumed, so any excess might be wasted.

Reply:
Yes this calculation gets us to maximum efficiency. You can also play with the calculator (https://gastoken.io/) to confirm.

```
gasUsed = 2,000,000
value = 48 =  (gasUsed + 14154) / 41130 

SELFDESTRUCT = -24000
SELFDESTRUCT_COST = 700 + 5000
refund = -878,400 = SELFDESTRUCT * 48 + SELFDESTRUCT_COST * 48

newGasUsed = 1,121,600 > gasUsed * 0.5
```